Allow Comments to be created with a passed `author_ip`

rachelbaker opened the pull request:
This PR allows the `comment_author_IP` value to be set when creating a comment.  While working on this, I remembered why we don't handle this field...  while with `wp_insert_comment()` you can as of WP 4.3 set the `comment_author_IP` **BUT** you cannot modify this field in `wp_update_comment()`.  
### Questions:
1. Do we allow this property to be set when creating a Comment, BUT not allow it to be edited?  How should that be reflected in the schema?
2. Should we add a filter for the default `comment_author_IP` value (currently `127.0.0.1`)?
## UPDATE:

This PR has been refreshed and as of WP 4.5 the `comment_author_ip` property can be modified.  View details on the refresh: https://github.com/WP-API/WP-API/pull/1880#issuecomment-217732303

Fixes #1788

rachelbaker committed:
Allow the `comment_author_IP` value to be set

rachelbaker committed:
Move the comment_author_IP validation to the `prepare_item_for_database` method and return error if invalid

rachelbaker committed:
Prevent fatal error when `prepare_item_for_database` returns a WP_Error

rachelbaker committed:
Add unit tests for creating a comment with `author_ip`

rachelbaker commented:
@WP-API/amigos thoughts?

danielbachhuber commented on lib/endpoints/class-wp-rest-comments-controller.php:
```diff
@@ -699,6 +707,14 @@ protected function prepare_item_for_database( $request ) {
 		}
 
+		if ( isset( $request['author_ip'] ) ) {
+			if ( filter_var( $request['author_ip'], FILTER_VALIDATE_IP, FILTER_FLAG_IPV4 ) ) {
```
What if we added `ip_address` as a new format in our schema, so validation could happen much earlier (and in batch)?

rachelbaker commented:
Related core ticket: https://core.trac.wordpress.org/ticket/35276

danielbachhuber commented:
Something I've been thinking about recently with regards to whether or not fields should be editable: import / export. If someone were to build a import / export workflow based on the REST API, which seems totally reasonable to me, it seems like a reasonable assumption they should be able to update most any fields for resources.

rachelbaker commented:
> Something I've been thinking about recently with regards to whether or not fields should be editable: import / export. If someone were to build a import / export workflow based on the REST API, which seems totally reasonable to me, it seems like a reasonable assumption they should be able to update most any fields for resources.

@danielbachhuber, can you please add those thoughts to https://core.trac.wordpress.org/ticket/35276?  Additional supporting opinions will help.

danielbachhuber committed:
Merge branch 'develop' into 1788-comment-author-IP

danielbachhuber committed:
Merge branch 'develop' into 1788-comment-author-IP

rachelbaker commented:
In WP 4.5 this will no longer be blocked on core.  I will pick this back up after release.

caglaroner commented:
@rachelbaker & @danielbachhuber  Great job on the WP-API and thank you very much for all your contributions. 

Our team is currently blocked on this issue (more specifically the post comment part) and we would hate to branch out and patch it on our end. 

Is there any timeline or progress on this issue?

rachelbaker committed:
Merge branch 'develop' into 1788-comment-author-IP

rachelbaker committed:
Remove read-only attribute from author_ip

rachelbaker committed:
Use string format type of 'ipv4' to validate comment author IP address

New function `rest_is_ip_address()` checks if a value is a valid v4 IP address. It uses regex because core cannot guarantee support for `filter_var`.

rachelbaker committed:
Use the schema to set `127.0.0.1` as the default comment author IP

rachelbaker committed:
Update tests and improve coverage

rachelbaker committed:
Add space to fix PHPCS error

rachelbaker committed:
Remove unneeded check on prepare_item_for_database returning a WP_Error on update

rachelbaker committed:
Similar to d051e65 this removes the unneeded check on comment create

rachelbaker commented:
@WP-API/amigos This PR has been refreshed and is ready for review.
- Handles setting and updating the `author_ip` property on comment create & update
- Uses 127.0.0.1 as the default `author_ip` value in the schema
- Adds a new `ipv4` string format (which matches the following example on the JSON Schema site: http://json-schema.org/example2.html) 
- Adds the IP validation utility function `rest_is_ip_address()` which ideally I would like to get into WP core.
